include/algorithms/public/SKMeans.hpp

```diff
@@ -0,0 +1,121 @@
+/*
+Part of the Fluid Corpus Manipulation Project (http://www.flucoma.org/)
+Copyright 2017-2019 University of Huddersfield.
+Licensed under the BSD-3 License.
+See license.md file in the project root for full license information.
+This project has received funding from the European Research Council (ERC)
+under the European Union’s Horizon 2020 research and innovation programme
+(grant agreement No 725899).
+*/
+
+#pragma once
+
+#include "../public/KMeans.hpp"
+#include "../util/FluidEigenMappings.hpp"
+#include "../../data/FluidDataSet.hpp"
+#include "../../data/FluidIndex.hpp"
+#include "../../data/FluidTensor.hpp"
+#include "../../data/TensorTypes.hpp"
+#include <Eigen/Core>
+#include <queue>
+#include <string>
+
+namespace fluid {
+namespace algorithm {
+
+class SKMeans : public KMeans
+{
+
+public:
+  void train(const FluidDataSet<std::string, double, 1>& dataset, index k,
+             index maxIter)
+  {
+    using namespace Eigen;
+    using namespace _impl;
+    assert(!mTrained || (dataset.pointSize() == mDims && mK == k));
+    MatrixXd dataPoints = asEigen<Matrix>(dataset.getData());
+    MatrixXd dataPointsT = dataPoints.transpose();
+    if (mTrained) { mAssignments = assignClusters(dataPointsT);}
+    else
+    {
+      mK = k;
+      mDims = dataset.pointSize();
+      initMeans(dataPoints);
+    }
+
+    while (maxIter-- > 0)
+    {
+      mEmbedding = mMeans.matrix() * dataPointsT;
+      auto assignments = assignClusters(mEmbedding);
+      if (!changed(assignments)) { break; }
+      else
+        mAssignments = assignments;
+      updateEmbedding();
+      computeMeans(dataPoints);
+    }
+    mTrained = true;
+  }
+
+
+  void transform(RealMatrixView data, RealMatrixView out,
+                 double alpha = 0.25) const
+  {
+    using namespace Eigen;
+    MatrixXd points = _impl::asEigen<Matrix>(data).transpose();
+    MatrixXd embedding = (mMeans.matrix() * points).array() - alpha;
+    embedding = (embedding.array() > 0).select(embedding, 0).transpose();
+    out = _impl::asFluid(embedding);
```

Comment:
```suggestion
    out <<= _impl::asFluid(embedding);
```
New copy assignment operator for `FluidTensorView`